byway/src/main/java/com/google/sps/data/Trip.java

```diff
@@ -0,0 +1,88 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+import com.google.appengine.api.datastore.Entity;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import java.util.Collection;
+import java.util.ArrayList;
+
+/**
+ * A class to make a Trip type, containing
+ * a specific id, destinations, interests and routes 
+ * to be used when updating a trip.
+ */
+public final class Trip {
+
+  private final String id;
+  private final String start;
+  private final ArrayList<String> destinations;
+  private final ArrayList<String> interests;
+  private final ArrayList<String> route;
+  public static final String DATASTORE_ENTITY_KIND = "Trip";
+
+  public Trip(String id, String start, Collection<String> destinations,
+              Collection<String> interests, Collection<String> route) {
+    this.id = checkNotNull(id, "id");
+    this.start = checkNotNull(start, "start");
+    Collection<String> validDestinations = checkNotNull(destinations, "destinations");
+    Collection<String> validInterests = checkNotNull(interests, "interests");
+    Collection<String> validRoute = checkNotNull(route, "route");
```

Comment:
I added checkNotNull  to see if the params passed in themselves were null. I thought it was overkill, but wasn't sure when/where to check for null values, so did it for the collections as well. If I removed this, would I just do
`this.interests = new ArrayList<>(interests)`? It wouldn't check if the collection is null, which would make a new arraylist but with null value(s).

Reply:
You can just call `checkNotNull` without doing anything with the returned value, the returned value doesn't matter. The function works roughly like this:

```java
static <T> T checkNotNull(T value, String failureMessage) {
  if (value == null) {
    throw new NullPointerException(failureMessage);
  }
  return value;
}
```

so if the value is null, you are guaranteed an exception, and the next line will not execute. If the value is not null, it is returned unmodified. Returning the same value unmodified is just a convenience feature which is there to allow you to write a null check in a single line when that's appropriate, like with the `start` and `id` where you write `this.id = checkNotNull(id, "id");`. when it's not appropriate to write the check in a single line, you can write `checkNotNull(destinations, "destinations");` on one line without using the returned value, and just assume that the value will not be null from then on (because if it was, you would have thrown an exception.